Fixes #5554 - fixed hostgroup and fqdn validations in multiple matchers

orrabin opened the pull request.

orrabin committed:
Fixes #5554 - fixed hostgroup and fqdn validations in multiple matchers

unorthodoxgeek commented:
review is cool, but let's add an automatic test to this.
something akin:

``` ruby
#in OverrideValueTest

test "it doesn't care about order" do
  override_value = OverrideValue.new(match: "hostgroup=Servers,osfamilyRedHat")
  assert override_value.match(whatever) #pseudo code
  reverse_override_value = OverrideValue.new(match: "hostgroup=Servers,osfamily=RedHat")
  assert reverse_override_value.match(whatever) #pseudo code
end
```

please make sure the test fails (i.e. run the pervious code, fail the test and then this test should pass)

orrabin commented:
@unorthodoxgeek test was added to another related PR: https://github.com/theforeman/foreman/pull/1925/files#diff-2dfd7aec89c8173edacdb4780e94a5f6R586
I'll move it to here and rebase the other PR when this is merged

unorthodoxgeek commented:
my name is Tom and I approve this PR :+1: 

i.e. @ohadlevy - merge plz :)

orrabin commented:
[test]

orrabin commented on test/unit/orchestration/dhcp_test.rb:
```diff
@@ -144,7 +144,7 @@ def test_bmc_should_have_valid_dhcp_record
 
   test "when an existing host change multiple attributes, both his dhcp and bmc dhcp records should be updated" do
-    h = FactoryGirl.create(:host, :with_dhcp_orchestration)
+    h = FactoryGirl.create(:host, :with_dhcp_orchestration, :mac => "aa:aa:ad:ab:bb:cc")
```
In line 153 `h.mac = h.mac.succ` when h.mac ends with 2f it is changed to 2g and becomes invalid.
Somehow the test I added for this PR caused the sequence of macs to end with 2f for the host in this test and so the test failed.
Setting the host to a different mac solves the problem.
Another option is not using succ or checking if the mac becomes invalid and then changing it again

dLobatog commented on test/unit/classification_test.rb:
```diff
@@ -3,7 +3,9 @@
 class ClassificationTest < ActiveSupport::TestCase
 
   def setup
+    hostgroup = FactoryGirl.create(:hostgroup)
```
This is creating a new hostgroup in every test, can you move it to only the relevant test (I think it should be only the  one introduced in this PR) and update the host with that host group there?

dLobatog commented on test/unit/classification_test.rb:
```diff
@@ -593,6 +595,43 @@ def setup
+    hostgroup.update_attributes(:parent => parent_hostgroup)
+
+    as_admin do
+      LookupValue.create! :lookup_key_id => key.id,
```
Just making sure, can these values be created through FactoryGirl or stubbed instead? All we really care about is testing the lookup_key does the proper match with the match regex & value, the associations and the rest are unneeded I think.

orrabin replied:
@eLobato I don't really understand what's unneeded. I can try to move this to FactoryGirl but the only associations I see are to lookup_key which is needed and to puppetclasses and environments so the classification will find these lookup_values and use them.

dLobatog commented:
:+1: for the fixes (although the duplication we have in lookup_key with ensure_xxxx_exists makes me cringe a bit), just 2 minor issues with the tests. 
Fixing the setup would be enough probably, although not hitting the db in the whole test would be awesome. Thanks @orrabin !

orrabin commented:
@eLobato I rebased and fixed your comments. If you approve the test I will open a refactoring issue to fix the existing tests too. 
I don't like the two `ensure` there but hostgroup has an extra test and the error message is not exactly the same so I can merge them but I'm not sure it's worth it for 3 lines.

orrabin commented:
@eLobato ping? I rebased again

dLobatog commented:
Merged as cdf75159a87d1fe0eb3d8a2595cfa46f68d8fa24, thanks @orrabin!
